include/modem/lte_lc.h

```diff
@@ -42,38 +52,111 @@ enum lte_lc_system_mode {
 	LTE_LC_FUNC_MODE_NORMAL			= 1,
 	LTE_LC_FUNC_MODE_OFFLINE		= 4,
-	LTE_LC_FUNC_MODE_OFFLINE_UICC_ON	= 44
+	LTE_LC_FUNC_MODE_DEACTIVATE_LTE		= 20,
+	LTE_LC_FUNC_MODE_ACTIVATE_LTE		= 21,
+	LTE_LC_FUNC_MODE_DEACTIVATE_GNSS	= 30,
+	LTE_LC_FUNC_MODE_ACTIVATE_GNSS		= 31,
+	LTE_LC_FUNC_MODE_OFFLINE_UICC_ON	= 44,
+};
+
+enum lte_lc_evt_type {
+	LTE_LC_EVT_NW_REG_STATUS,
+	LTE_LC_EVT_PSM_UPDATE,
+	LTE_LC_EVT_EDRX_UPDATE,
+	LTE_LC_EVT_RRC_UPDATE,
+	LTE_LC_EVT_CELL_UPDATE,
+};
+
+enum lte_lc_rrc_mode {
+	LTE_LC_RRC_MODE_IDLE,
+	LTE_LC_RRC_MODE_CONNECTED,
+};
+
+struct lte_lc_psm_cfg {
+	int tau;		/* Periodic Tracking Area Update interval */
+	int active_time;	/* Active-time (time from RRC idle to PSM) */
+};
+
+struct lte_lc_edrx_cfg {
+	float edrx;	/* eDRX interval value [s] */
+	float ptw;	/* Paging time window [s] */
+};
+
+struct lte_lc_cell {
+	u32_t id;	/* E-UTRAN cell ID */
+	u32_t tac;	/* Tracking Area Code */
+};
+
+struct lte_lc_evt {
+	enum lte_lc_evt_type type;
+	union {
+		enum lte_lc_nw_reg_status nw_reg_status;
+		enum lte_lc_rrc_mode rrc_mode;
+		struct lte_lc_psm_cfg psm_cfg;
+		struct lte_lc_edrx_cfg edrx_cfg;
+		struct lte_lc_cell cell;
+	};
 };
 
-/** @brief Function for initializing
- * the modem.  NOTE: a follow-up call to lte_lc_connect()
- * must be made.
+typedef void(*lte_lc_evt_handler_t)(const struct lte_lc_evt *const evt);
+
+/** @brief Register event handler for LTE events.
+ *
+ *  @param handler Event handler. Handler is de-registered if parameter is
+ *		   NULL.
+ */
+void lte_lc_register_handler(lte_lc_evt_handler_t handler);
+
+/** @brief Function for initializing the modem.
+ *
+ * @note a follow-up call to lte_lc_connect() must be made to establish an LTE
+ *	 connection.
  *
  * @return Zero on success or (negative) error code otherwise.
  */
 int lte_lc_init(void);
 
 /** @brief Function to make a connection with the modem.
- * NOTE: prior to calling this function a call to lte_lc_init()
- * must be made.
+ * @note prior to calling this function a call to @ref lte_lc_init
+ *	 must be made.
  *
  * @return Zero on success or (negative) error code otherwise.
  */
 int lte_lc_connect(void);
 
-/** @brief Function for initializing
- * and make a connection with the modem
+/** @brief Function for initializing and make a connection with the modem
  *
  * @return Zero on success or (negative) error code otherwise.
  */
 int lte_lc_init_and_connect(void);
 
+/**@brief Connect to LTE network. Non-blocking.
+ *
+ * @note @ref lte_lc_init must be called before this function.
+ *
+ * @param handler Event handler for receiving LTE events.
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+int lte_lc_connect_async(lte_lc_evt_handler_t handler);
+
+/**@brief Initializes the LTE module, configures the modem and connects to LTE
+ *	  network. Non-blocking.
+ *
+ * @param handler Event handler for receiving LTE events.
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+
```

Comment:
Extra blank line.
Could you please document the `EINVAL` error code returned? The handled parameter can be NULL but there is no information about this.

Reply:
Added that the parameter can be `NULL` if handler is already registered.
Not sure about adding `@retval -EINVAL` for the same reasons as above. Many error codes from underlying modules are forwarded, and not sure how to handle that.

Solved like this for now:
```
/**@brief Connect to LTE network. Non-blocking.
 *
 * @note @ref lte_lc_init must be called before this function.
 *
 * @param handler Event handler for receiving LTE events. The parameter can be
 *		  NULL if an event handler is already registered.
 *
 * @return Zero on success, -EINVAL if no handler is provided and not already
 *	   registered, otherwise a (negative) error code.
```